2021/day14/day14.java

```diff
@@ -0,0 +1,111 @@
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.io.File;
+
+public class day14 {
+    public static class Rule {
+        String pair;
+        String outcome;
+    }
+
+    public static String readTemplate(String filename) {
+        String template = null;
+        Scanner sc = null;
+        try {
+            sc = new Scanner(new File(filename));
+            template = sc.next();
+        } catch (Exception e) {
+            System.out.println("Something went wrong" + e);
+        } finally {
+            sc.close();
+        }
+        return template;
+    }
+
+        /*
+         * Given the name of a file containing a list of rules for creating
+         * new elements in the form of arrow-delimited strings, e.g. "CH -> B",
+         * read and transform them into a list of `Rule`s
+         */
+    public static ArrayList<Rule> readRules(String filename) {
+        Scanner sc = null;
+        ArrayList<Rule> input = new ArrayList<>();
+        try {
+            sc = new Scanner(new File(filename));
+            sc.useDelimiter("\\n| -> ");
+            while (sc.hasNext()) {
+                Rule rule = new Rule();
+                rule.pair = sc.next();
+                rule.outcome = sc.next();
+                input.add(rule);
+            }
+        } catch (Exception e) {
+            System.out.println("Something went wrong" + e);
+        } finally {
+            sc.close();
+        }
+        return input;
+    }
+
+    /*
+     * Split the template into two-letter chunks and count
+     * the occurrences of them in the template.
+     *
+     * E.g. given a template
+     * NCCNC
+     * Transform it into a map
+     * {"NC": 2, "CN": 1, "CC": 1}
+     */
+    public static HashMap<String, Long> splitToChunks(String template) {
+        HashMap<String, Long> chunks = new HashMap<>();
+        for (int i = 0; i < template.length() - 1; i++) {
+            chunks.merge(template.substring(i, i + 2), 1L, Long::sum);
+        }
+        return chunks;
+    }
+
+    /*
+     * Split the template into individual letters and count the occurrences
+     * of them in the template.
+     *
+     * E.g. given a template
+     * NCNBCHB
+     * Transform it into a map
+     * {"N":2, "C":2, "B":2, "H":1}
+     */
```

Comment:
Try to keep it in the bounds of 80 characters per line from now on. This is especially easy in the documentation where you don't have to care about how it looks. In VSCode you can probably even turn on auto wrap on column 80 or something.

Reply:
![Screenshot from 2021-12-20 18-26-45](https://user-images.githubusercontent.com/87368857/146807751-402b6766-3fb4-4a66-8889-c5d01ef114d3.png)
It is within 80 on my screen with the exception of 95-104, but I will have a look at auto wrap and keep it in mind in the future.